cli/tools/jupyter.rs

```diff
@@ -0,0 +1,893 @@
+    let metadata_bytes = zmq_msg.get(3).unwrap();
+    let content_bytes = zmq_msg.get(4).unwrap();
+
+    hmac_verify(
+      hmac_key,
+      expected_signature_bytes,
+      header_bytes,
+      parent_header_bytes,
+      metadata_bytes,
+      content_bytes,
+    )?;
+
+    // TODO(bartomieju): can these unwraps be better handled?
+    let header: MessageHeader = serde_json::from_slice(header_bytes).unwrap();
+    let parent_header: MessageHeader =
+      serde_json::from_slice(parent_header_bytes).unwrap();
+    let metadata: Value = serde_json::from_slice(metadata_bytes).unwrap();
+    let content: Value = serde_json::from_slice(content_bytes).unwrap();
+
+    let msg = Message {
+      is_reply: false,
+      r#type: header.msg_type.clone(),
+      header,
+      parent_header: Some(parent_header),
+      metadata,
+      content,
+      // FIXME:
+      session_id: "".to_string(),
+    };
+
+    Ok(msg)
+  }
+}
+
+struct PubComm {
+  conn_str: String,
+  session_id: String,
+  hmac_key: hmac::Key,
+  socket: zeromq::PubSocket,
+}
+
+impl PubComm {
+  pub fn new(
+    conn_str: String,
+    session_id: String,
+    hmac_key: hmac::Key,
+  ) -> Self {
+    println!("iopub connection: {}", conn_str);
+    Self {
+      conn_str,
+      session_id,
+      hmac_key,
+      socket: zeromq::PubSocket::new(),
+    }
+  }
+
+  pub async fn connect(&mut self) -> Result<(), AnyError> {
+    self.socket.bind(&self.conn_str).await?;
+
+    Ok(())
+  }
+
+  pub async fn send(&mut self, msg: Message) -> Result<(), AnyError> {
+    let zmq_msg = msg.serialize(&self.hmac_key);
+    self.socket.send(zmq_msg).await?;
+    Ok(())
+  }
+}
+
+struct DealerComm {
+  name: String,
+  conn_str: String,
+  session_id: String,
+  hmac_key: hmac::Key,
+  socket: zeromq::DealerSocket,
+}
+
+impl DealerComm {
+  pub fn new(
+    name: &str,
+    conn_str: String,
+    session_id: String,
+    hmac_key: hmac::Key,
+  ) -> Self {
+    println!("dealer '{}' connection: {}", name, conn_str);
+    Self {
+      name: name.to_string(),
+      conn_str,
+      session_id,
+      hmac_key,
+      socket: zeromq::DealerSocket::new(),
+    }
+  }
+
+  pub async fn connect(&mut self) -> Result<(), AnyError> {
+    self.socket.bind(&self.conn_str).await?;
+
+    Ok(())
+  }
+
+  pub async fn recv(&mut self) -> Result<ZmqMessage, AnyError> {
+    let msg = self.socket.recv().await?;
+    Ok(msg)
+  }
+
+  pub async fn send(&mut self, msg: Message) -> Result<(), AnyError> {
+    let zmq_msg = msg.serialize(&self.hmac_key);
+    self.socket.send(zmq_msg).await?;
+    Ok(())
+  }
+}
+
+async fn create_zmq_reply(name: &str, conn_str: &str) -> Result<(), AnyError> {
+  println!("reply '{}' connection string: {}", name, conn_str);
+
+  let mut sock = zeromq::RepSocket::new(); // TODO(apowers313) exact same as dealer, refactor
+  sock.monitor();
+  sock.bind(conn_str).await?;
+
+  loop {
+    let msg = sock.recv().await?;
+    dbg!(&msg);
+    println!("{} got packet!", name);
+  }
```

Comment:
Double-checking: this recv-and-drop is intentional?

Aside: `dbg!()` _and_ `println!()` is kind of excessive.

Reply:
Still WIP